paddle/framework/net.h

```diff
@@ -123,48 +91,27 @@ class PlainNet : public Net {
    * scope will be used instead. If no OpContext is provicded, default context
    * will be used.
    */
-  virtual void Run(std::shared_ptr<Scope> scope, DeviceContext *ctx) override;
+  void Run(const std::shared_ptr<Scope>& scope,
+           const platform::DeviceContext& dev_ctx) const override;
 
   /**
    * @brief Add an operator to this network.
    */
-  virtual OpIndex AddOp(const OpProto &def) override;
+  void AddOp(const OpDesc& def) override;
 
   /**
    * @brief Add all optimizer operators related into the network.
    */
-  virtual void AddOptimizerOps(const OpAttrs &attrs) override;
+  void AddOptimizerOps() override {}
 
   /**
    * @brief Add all backward operators related into the network.
    */
-  virtual void AddBackwardOps() override;
-
-  virtual ~PlainNet() override {}
-
- protected:
-  /**
-   * @brief Build the network.
-   *
-   * Create operators accordding to `def`, will be called by the constructor.
-   */
-  void BuildNet(const NetDesc &def);
-
-  /**
-   * @brief Add an operator into this network.
-   *
-   * Add a operator which is identified as `type` and has attributes described
-   * in `attrs`, the `inputs` are the keys of readonly input variables,
-   * `outputs` are keys of mutable output variables. An `OpIndex` will be
-   * returned to indicate the offset of the new operator in `ops_`.
-   */
-  OpIndex AddOp(const std::string &type, const std::vector<std::string> &inputs,
-                const std::vector<std::string> &outputs,
-                const OpAttrs &attrs = OpAttrs());
+  void AddBackwardOps() override {}
 
  private:
   // the operators owned by `Network`.
-  std::vector<Operator> ops_;
+  std::vector<std::unique_ptr<OperatorBase>> ops_;
```

Comment:
std::shared_ptr<OperatorBase> is better.

The `std::shared_ptr` is the only pointer can expose to Python correctly.